src/realm/table.cpp

```diff
@@ -2579,12 +2599,33 @@ Obj Table::create_object(GlobalKey object_id, const FieldValues& values)
         throw LogicError(LogicError::wrong_kind_of_table);
     ObjKey key = object_id.get_local_key(get_sync_file_id());
 
+    // Check if the object already exists.
+    if (is_valid(key)) {
```

Comment:
For performance reasons we should avoid is_valid().... get_object().... both are traversing the tree searching for the key.

Reply:
That makes sense. Is there a way to call `get_object()` while being unsure that the key exists?

Reply:
try .. catch()...

Reply:
That seems like the entirely wrong thing to do here. We really need to rethink some of the abstractions on Table, it's getting a bit bad. I'll do what you suggest, accompanied by a FIXME. :-)

Reply:
I was just referring to what the rest of the API would support. But you are free to provide a different interface. There is a try_get() on the ClusterNode, you can introduce a nice API above that if you prefer. I don't know how low the overhead of option or variant is - guess they are the obvious alternative to exceptions.

Reply:
Or I can...?

Reply:
try_get() on Table would be nice to have.

Reply:
@finnschiermer In general, I believe that `catch` blocks should not exist anywhere in Core (outside of tests). I am much less worried about the performance of an extra boolean + branches in the return channel on the happy path, than I am about the performance of stack unwinding on the less-than-happy path.

And it makes running Core tests in a debugger super annoying too. :-)

Reply:
Thinking more about it, I can't actually use exceptions in this way here, at least not without rewrangling a couple of things, because the other `create_object()` functions are idempotent (including `create_object(GlobalKey)`), and we still need to check for tombstones.

`create_object(ObjKey)` is the odd one out, because it is the only one that isn't idempotent (and the only reason that it isn't is that there are unit tests relying on exceptions being thrown from it on the "happy path", which is a problem on its own).

Note that `create_object(non-null ObjKey)` should basically never happen in practice. The other `create_object(...)` functions go through the `create_object_internal()` function, which does no bookkeeping.